freqtrade/freqai/data_kitchen.py

```diff
@@ -0,0 +1,1002 @@
+        )
+
+        config_path = Path(self.config["config_files"][0])
+
+        if not self.full_path.is_dir():
+            self.full_path.mkdir(parents=True, exist_ok=True)
+            shutil.copy(
+                config_path.resolve(),
+                Path(self.full_path / config_path.parts[-1]),
+            )
+
+        return full_timerange
+
+    def check_if_new_training_required(self, trained_timestamp: int) -> Tuple[bool, TimeRange]:
+
+        time = datetime.datetime.now(tz=datetime.timezone.utc).timestamp()
+        trained_timerange = TimeRange()
+        if trained_timestamp != 0:
+            elapsed_time = (time - trained_timestamp) / SECONDS_IN_DAY
+            retrain = elapsed_time > self.freqai_config.get('backtest_period')
+            if retrain:
+                trained_timerange.startts = int(time - self.freqai_config.get(
+                                             'train_period', 0) * SECONDS_IN_DAY)
+                trained_timerange.stopts = int(time)
+        else:  # user passed no live_trained_timerange in config
+            trained_timerange.startts = int(time - self.freqai_config.get('train_period') *
+                                            SECONDS_IN_DAY)
+            trained_timerange.stopts = int(time)
+            retrain = True
+
+        # if retrain:
+        #     coin, _ = metadata['pair'].split("/")
+        #     # set the new data_path
+        #     self.data_path = Path(self.full_path / str("sub-train" + "-" +
+        #                            str(int(trained_timerange.stopts))))
+
+        #     self.model_filename = "cb_" + coin.lower() + "_" + str(int(trained_timerange.stopts))
+        #     # this is not persistent at the moment TODO
+        #     self.freqai_config['live_trained_timerange'] = str(int(trained_timerange.stopts))
+        #     # enables persistence, but not fully implemented into save/load data yer
+        #     self.data['live_trained_timerange'] = str(int(trained_timerange.stopts))
+
+        return retrain, trained_timerange
+
+    def set_new_model_names(self, metadata: dict, trained_timerange: TimeRange):
+
+        coin, _ = metadata['pair'].split("/")
+        # set the new data_path
+        self.data_path = Path(self.full_path / str("sub-train" + "-" +
+                              metadata['pair'].split("/")[0] +
+                              str(int(trained_timerange.stopts))))
+
+        self.model_filename = "cb_" + coin.lower() + "_" + str(int(trained_timerange.stopts))
+
+        # self.freqai_config['live_trained_timerange'] = str(int(trained_timerange.stopts))
+        # enables persistence, but not fully implemented into save/load data yer
+        # self.data['live_trained_timerange'] = str(int(trained_timerange.stopts))
+
+    def download_new_data_for_retraining(self, timerange: TimeRange, metadata: dict) -> None:
+
+        exchange = ExchangeResolver.load_exchange(self.config['exchange']['name'],
```

Comment:
I assume this is for the live refresh?

Why are you creating a new instance here instead of using the one that exists already?
by creating a new instance, you're creating problems with rate-limiting (it's now a new, separate rate-limiter) - and with that potentially problems for both instances - as they might refresh both (assuming your refresh here goes "across" a new candle time - then both would reload data at the same time - very likely locking you out (depending on the exchange, but still a very real possibility).

the below used method (`refresh_backtest_ohlcv_data()`)  was due for removal, and was only "left" for edge. Now i don't mind having this used here - but Edge does not recreate a new exchange class - so the regular rate-limit stuff can still fly onwards.

Reply:
The reason I instantiated a dummy exchange was becaue I didn't find a way to access FreqtradeBot.exchange without passing it to the strategy and then into FreqAI. Is there an easier way to gain access to FreqtradeBot.exchange from inside the strategy?

Reply:
scratch that, I can use the dataprovider like this:

strategy.dp._exchange to grab it.

Reply:
After further private discussion, we concluded that the cleanest solution is to restructure the handling of training data in dry/live background threads. This is a relatively intensive rehaul, so it has been added to the wishlist.

However, currently we are not observing any ratelimiting issues associated with pinging the exchange on the background thread (in parallel with the foreground bot loop). So we will keep the existing "hack" until we have time to restructure the data handling.

Resolving this thread accordingly.